talismane_core/src/main/java/com/joliciel/talismane/tokeniser/TokenSequence.java

```diff
@@ -28,54 +29,97 @@
 /**
  * A sequence of tokens. Note: by default, List iteration and associated methods
  * will only return non-whitespace tokens. For a list that includes whitespace
- * tokens, use the listWithWhiteSpace() method.
+ * tokens, use the listWithWhiteSpace() method.<br/>
+ * <br/>
+ * Only a single attribute of a given key can be added per token. If two
+ * attribute placeholders overlap and assign the same key, the one with the
+ * lower start index is privileged for all tokens, and the other one is skipped
+ * for all tokens. If they have the same start index, the one with the higher
+ * end index is priveleged for all tokens, and the other one is skipped for all
+ * tokens.
  * 
  * @author Assaf Urieli
  *
  */
-public class TokenSequence extends ArrayList<Token> implements Serializable {
+public class TokenSequence extends ArrayList<Token>implements Serializable {
 	private static final Logger LOG = LoggerFactory.getLogger(TokenSequence.class);
 	private static final long serialVersionUID = 1L;
 
 	private List<Integer> tokenSplits;
-	private String text;
-	private Sentence sentence;
-	private List<Token> listWithWhiteSpace = new ArrayList<Token>();
+	private final Sentence sentence;
+	private final List<Token> listWithWhiteSpace;
 	private List<Token> tokensAdded = null;
 	private double score = 0;
 	private boolean scoreCalculated = false;
-	TokenisedAtomicTokenSequence underlyingAtomicTokenSequence;
+	private TokenisedAtomicTokenSequence underlyingAtomicTokenSequence;
 	private Integer atomicTokenCount = null;
 	private boolean finalised = false;
 	private boolean sentenceTagsAdded = false;
 	private boolean withRoot = false;
 	private PosTagSequence posTagSequence;
-	protected boolean textProvided = false;
+	private final Map<Integer, Annotation<TokenPlaceholder>> placeholderMap;
+	@SuppressWarnings("rawtypes")
```

Comment:
To make it parametrised, we'd need to add a wildcard to TokenAttribute:
`private final Map<Integer, Annotation<TokenAttribute<?>>`
However, in the AnnotatedText method
`public <T> List<Annotation<T>> getAnnotations(Class<T> clazz)`
I'd somehow need to be able to return a `List<Annotation<TokenAttribute<?>>>` or to cast the items in the list from Annotation<TokenAttribute> to Annotation<TokenAttribute<?>>, which Java refuses to do (Type Mismatch).